lib/internal/util.js

```diff
@@ -187,3 +191,58 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    return orig[kCustomPromisifiedSymbol];
+  }
+
+  if (isAsyncFunction(orig)) {
+    return orig;
+  }
+
+  // Names to create an object from in case the callback receives multiple
+  // arguments, e.g. ['stdout', 'stderr'] for child_process.exec.
+  const argumentNames = orig[kCustomPromisifyArgsSymbol];
+
+  const fn = function(...args) {
+    const promise = createPromise();
+    try {
+      orig.call(this, ...args, (err, ...values) => {
+        if (err) {
+          promiseReject(promise, err);
+        } else if (values.length === 1) {
+          promiseResolve(promise, values[0]);
+        } else if (argumentNames !== undefined) {
+          const obj = {};
+          for (var i = 0; i < argumentNames.length; i++)
+            obj[argumentNames[i]] = values[i];
+          promiseResolve(promise, obj);
```

Comment:
The logic should probably prefer `argumentNames !== undefined` over `values.length === 1`, since if for some reason the second argument (like `stderr`) is undefined we would still want to resolve an object.